packages/grafana-runtime/src/services/LocationService.ts

```diff
@@ -0,0 +1,155 @@
+import { UrlQueryMap, urlUtil } from '@grafana/data';
+import * as H from 'history';
+import { LocationUpdate } from './LocationSrv';
+import { createLogger } from '@grafana/ui';
+import { config } from '../config';
+
+/**
+ * @alpha
+ * A wrapper to help work with browser location and history
+ */
+export interface LocationService {
+  partial: (query: Record<string, any>, replace?: boolean) => void;
+  push: (location: H.Path | H.LocationDescriptor<any>) => void;
+  replace: (location: H.Path | H.LocationDescriptor<any>, forceRouteReload?: boolean) => void;
+  reload: () => void;
+  getLocation: () => H.Location;
+  getHistory: () => H.History;
+  getSearch: () => URLSearchParams;
+  getSearchObject: () => UrlQueryMap;
+
+  /**
+   * This is from the old LocationSrv interface
+   * @deprecated use partial, push or replace instead */
+  update: (update: LocationUpdate) => void;
+}
+
+class HistoryWrapper implements LocationService {
+  private readonly history: H.History;
+
+  constructor(history?: H.History) {
+    // If no history passed create an in memory one if being called from test
+    this.history =
+      history || process.env.NODE_ENV === 'test'
+        ? H.createMemoryHistory({ initialEntries: ['/'] })
+        : H.createBrowserHistory({ basename: config.appSubUrl ?? '/' });
+
+    // For debugging purposes the location service is attached to global _debug variable
+    if (process.env.NODE_ENV !== 'production') {
+      // @ts-ignore
+      let debugGlobal = window['_debug'];
+      if (debugGlobal) {
+        debugGlobal = {
+          ...debugGlobal,
+          location: this,
+        };
+      } else {
+        debugGlobal = {
+          location: this,
+        };
+      }
+      // @ts-ignore
+      window['_debug'] = debugGlobal;
+    }
+
+    this.partial = this.partial.bind(this);
```

Comment:
This seems likes the most optimal way to bind class functions, using arrow functions create 100% new functions in constructor, But mainly it's because I hate mixing arrow functions and norma prototypee functions, I find it just makes the code a lot messier when I see a class with some prototype functions and some arrow instance functions. So moved this to using this method to keep all functions looking the same. But I actually think we can remove these binds, don't think they are needed

Reply:
I think performance-wise it's almost the same since `bind` also creates new functions. But my main concern is that if there's a lot of the methods bound like this, it will make code messier and harder to maintain (e.g. if method name changes, it now needs to be updated here as well). Another benefit of arrow methods is proper `this` scoping.